File: typescript/src/clients/receipts.ts
import { Transport } from '../transport.js';

/** Client for AI-powered receipt and ticket data extraction. */
export class ReceiptsClient {
  constructor(private readonly transport: Transport) {}

  /** Extract structured data from a receipt image or PDF. Requires AI processing consent in portal settings. */
  async extract(file: Buffer, fileName: string): Promise<ReceiptExtractionResult> {
    return this.transport.upload<ReceiptExtractionResult>('/api/receipts/extract', 'file', file, fileName);
  }

  /** Extract data from multiple receipts sequentially. Convenience wrapper around extract(). */
  async extractBatch(files: Array<{ buffer: Buffer; name: string }>): Promise<BatchExtractionResult> {
    const results: BatchItemResult[] = [];
    let successCount = 0;
    for (let i = 0; i < files.length; i++) {
      try {
        const result = await this.extract(files[i].buffer, files[i].name);
        results.push({ index: i, filename: files[i].name, success: true, result });
        successCount++;
      } catch (e) {
        results.push({ index: i, filename: files[i].name, success: false, error: String(e) });
      }
    }
    return {
      totalFiles: files.length,
      successCount,
      failureCount: files.length - successCount,
      results,
    };
  }

  /** Submit receipt for async extraction with optional webhook callback. */
  async extractAsync(file: Buffer, fileName: string, options?: {
    callbackUrl?: string;
    callbackSecret?: string;
  }): Promise<ReceiptJob> {
    const extraFields: Record<string, string> = {};
    if (options?.callbackUrl) extraFields.callbackUrl = options.callbackUrl;
    if (options?.callbackSecret) extraFields.callbackSecret = options.callbackSecret;
    return this.transport.upload<ReceiptJob>('/api/receipts/extract-async', 'file', file, fileName, extraFields);
  }

  /** Extract receipt and generate an expense report document. */
  async toDocument(file: Buffer, fileName: string, options?: {
    format?: string;
    templateName?: string;
    title?: string;
  }): Promise<ReceiptToDocumentResult> {
    const extraFields: Record<string, string> = {};
    if (options?.format) extraFields.format = options.format;
    if (options?.templateName) extraFields.templateName = options.templateName;
    if (options?.title) extraFields.title = options.title;
    return this.transport.upload<ReceiptToDocumentResult>('/api/receipts/to-document', 'file', file, fileName, extraFields);
  }

  /** Export receipt data as CSV (DATEV-compatible). */
  async exportCsv(receipts: ReceiptData[]): Promise<Buffer> {
    return this.transport.requestBytes('POST', '/api/receipts/export/csv', receipts);
  }

  /** Export receipt data as Excel (XLSX). */
  async exportXlsx(receipts: ReceiptData[]): Promise<Buffer> {
    return this.transport.requestBytes('POST', '/api/receipts/export/xlsx', receipts);
  }

  /** Get async job status. */
  async getJob(jobId: string): Promise<ReceiptJob> {
    return this.transport.requestJson<ReceiptJob>('GET', `/api/receipts/jobs/${jobId}`);
  }

  /** Get async job result. */
  async getJobResult(jobId: string): Promise<ReceiptExtractionResult> {
    return this.transport.requestJson<ReceiptExtractionResult>('GET', `/api/receipts/jobs/${jobId}/result`);
  }

  /** List all async jobs. */
  async listJobs(): Promise<ReceiptJob[]> {
    return this.transport.requestJson<ReceiptJob[]>('GET', '/api/receipts/jobs');
  }
}

// -- Types -------------------------------------------------------------------

export interface VatBreakdown {
  rate: number;
  netAmount: number;
  vatAmount: number;
  grossAmount: number;
}

export interface LineItem {
  description: string;
  quantity?: number;
  unitPrice?: number;
  totalPrice?: number;
  vatRate?: number;
}

export interface ReceiptData {
  receiptType?: string;
  issuer?: string;
  date?: string;
  time?: string;
  totalAmount?: number;
  netAmount?: number;
  currency?: string;
  vatBreakdown?: VatBreakdown[];
  items?: LineItem[];
  paymentMethod?: string;
  referenceNumber?: string;
  tipAmount?: number;
  taxNumber?: string;
  vatId?: string;
  issuerAddress?: string;
  language?: string;
  country?: string;
  skr03Account?: string;
  skr03Description?: string;
  skr04Account?: string;
  skr04Description?: string;
  expenseCategory?: string;
  metadata?: Record<string, unknown>;
}

export interface ReceiptExtractionResult {
  receiptData: ReceiptData;
  missingFields?: string[];
  uncertainFields?: string[];
  warnings?: string[];
  qualityHints?: string[];
  inputType?: string;
  aiExtractionUsed?: boolean;
  confidence?: number;
  message?: string;
}

export interface BatchItemResult {
  index: number;
  filename: string;
  success: boolean;
  result?: ReceiptExtractionResult;
  error?: string;
}

export interface BatchExtractionResult {
  totalFiles: number;
  successCount: number;
  failureCount: number;
  results: BatchItemResult[];
}

export interface ReceiptJob {
  jobId: string;
  status: string;
  filename?: string;
  result?: ReceiptExtractionResult;
  error?: string;
  createdAt?: string;
  completedAt?: string;
  durationMs?: number;
}

export interface ReceiptToDocumentResult {
  extraction: ReceiptExtractionResult;
  generatedHtml?: string;
  suggestedFilename?: string;
}

File: python/src/docgen/clients/receipts.py
"""Client for AI-powered receipt and ticket data extraction."""

from __future__ import annotations

from typing import Any

from docgen._transport import Transport


class ReceiptsClient:
    """Client for receipt extraction, export, and async processing.

    Requires AI processing consent in portal settings (GDPR).
    """

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    # ── Extraction ───────────────────────────────────────────────────

    def extract(self, file_bytes: bytes, file_name: str, content_type: str = "image/jpeg") -> dict[str, Any]:
        """Extract structured data from a receipt image or PDF."""
        return self._t.upload("/api/receipts/extract",
            files={"file": (file_name, file_bytes, content_type)})

    def extract_batch(self, files: list[tuple[str, bytes, str]]) -> dict[str, Any]:
        """Extract data from multiple receipts sequentially.

        Args:
            files: List of (filename, bytes, content_type) tuples.
        """
        results = []
        success_count = 0
        for i, (name, data, ct) in enumerate(files):
            try:
                result = self.extract(data, name, ct)
                results.append({"index": i, "filename": name, "success": True, "result": result})
                success_count += 1
            except Exception as e:
                results.append({"index": i, "filename": name, "success": False, "error": str(e)})
        return {
            "totalFiles": len(files),
            "successCount": success_count,
            "failureCount": len(files) - success_count,
            "results": results,
        }

    # ── Async extraction ─────────────────────────────────────────────

    def extract_async(
        self,
        file_bytes: bytes,
        file_name: str,
        content_type: str = "image/jpeg",
        *,
        callback_url: str | None = None,
        callback_secret: str | None = None,
    ) -> dict[str, Any]:
        """Submit receipt for async extraction with optional webhook callback."""
        data: dict[str, str] = {}
        if callback_url:
            data["callbackUrl"] = callback_url
        if callback_secret:
            data["callbackSecret"] = callback_secret
        return self._t.upload("/api/receipts/extract-async",
            files={"file": (file_name, file_bytes, content_type)},
            data=data)

    # ── Document generation ──────────────────────────────────────────

    def to_document(
        self,
        file_bytes: bytes,
        file_name: str,
        content_type: str = "image/jpeg",
        *,
        format: str = "PDF",
        template_name: str | None = None,
        title: str = "Spesenbeleg",
    ) -> dict[str, Any]:
        """Extract receipt and generate expense report document."""
        data: dict[str, str] = {"format": format, "title": title}
        if template_name:
            data["templateName"] = template_name
        return self._t.upload("/api/receipts/to-document",
            files={"file": (file_name, file_bytes, content_type)},
            data=data)

    # ── Export ────────────────────────────────────────────────────────

    def export_csv(self, receipts: list[dict[str, Any]]) -> bytes:
        """Export receipt data as CSV (DATEV-compatible)."""
        return self._t.request_bytes("POST", "/api/receipts/export/csv", json=receipts)

    def export_xlsx(self, receipts: list[dict[str, Any]]) -> bytes:
        """Export receipt data as Excel (XLSX)."""
        return self._t.request_bytes("POST", "/api/receipts/export/xlsx", json=receipts)

    # ── Async jobs ───────────────────────────────────────────────────

    def get_job(self, job_id: str) -> dict[str, Any]:
        """Get async job status."""
        return self._t.request_json("GET", f"/api/receipts/jobs/{job_id}")

    def get_job_result(self, job_id: str) -> dict[str, Any]:
        """Get async job extraction result."""
        return self._t.request_json("GET", f"/api/receipts/jobs/{job_id}/result")

    def list_jobs(self) -> list[dict[str, Any]]:
        """List all async receipt jobs."""
        return self._t.request_list("GET", "/api/receipts/jobs")
